source/postprocess/visualization/theta_tauisa.cc

```diff
@@ -0,0 +1,215 @@
+                                                             new Vector<float>(this->get_triangulation().n_active_cells()));
+
+        const QMidpoint<dim> quadrature_formula;
+        const unsigned int n_q_points = quadrature_formula.size(); // this is 1 for QMidpoint
+
+        FEValues<dim> fe_values(this->get_mapping(), this->get_fe(),
+                                quadrature_formula,
+                                update_values | update_gradients | update_quadrature_points);
+
+        // Set up material models
+        MaterialModel::MaterialModelInputs<dim> in(n_q_points,
+                                                   this->n_compositional_fields());
+        MaterialModel::MaterialModelOutputs<dim> out(n_q_points,
+                                                     this->n_compositional_fields());
+
+        // Set up cell iterator for looping
+        typename DoFHandler<dim>::active_cell_iterator cell =
+          this->get_dof_handler().begin_active(), endc =
+            this->get_dof_handler().end();
+
+        // Loop over cells and calculate theta in each one
+        // Note that we start after timestep 0 because we need the strain rate,
+        // which doesn't exist during the initial step
+        unsigned int cell_index = 0;
+        for (; cell != endc; ++cell, ++cell_index)
+          {
+            if (cell->is_locally_owned() && this->get_timestep_number() > 0)
+              {
+
+                // Fill the material model objects for the cell (for strain rate)
+                fe_values.reinit(cell);
+                in.reinit(fe_values, cell, this->introspection(),
+                          this->get_solution(), true);
+                // Also get velocities and velocity gradients
+                std::vector<Tensor<1, dim> > velocities(n_q_points,
+                                                        Tensor<1, dim>());
+                std::vector<Tensor<2, dim> > velocity_gradient(n_q_points,
+                                                               Tensor<2, dim>());
+                fe_values[this->introspection().extractors.velocities].get_function_values(
+                  this->get_solution(), velocities);
+                fe_values[this->introspection().extractors.velocities].get_function_gradients(
+                  this->get_solution(), velocity_gradient);
+
+                // Calculate eigenvalues of strain rate and take maximum (absolute value)
+                // to get tauISA, the timescale for grain rotation toward the infinite strain axis
+                const SymmetricTensor<2, dim> strain_rate = in.strain_rate[0];
+                std::array<double, dim> strain_rate_eigenvalues = eigenvalues(
+                                                                    strain_rate);
+                double lamda1 = std::max(std::abs(strain_rate_eigenvalues[0]),
+                                         std::abs(strain_rate_eigenvalues[dim]));
```

Comment:
This looks like an invalid array access -- you are accessing element `dim`, but the last value index is `dim-1` (there are `dim` elements, starting at zero).

Reply:
In fact, it's not clear to me what you are doing here -- what do you do in 3d? assume that the largest eigenvalue is either the first or the last one in the array?